src/basic/terminal-util.c

```diff
@@ -1135,14 +1135,19 @@ int open_terminal_in_namespace(pid_t pid, const char *name, int mode) {
 }
 
 bool colors_enabled(void) {
-        const char *colors;
+        static int enabled = -1;
 
-        colors = getenv("SYSTEMD_COLORS");
-        if (!colors) {
-                if (streq_ptr(getenv("TERM"), "dumb"))
-                        return false;
-                return on_tty();
+        if (_unlikely_(enabled < 0)) {
```

Comment:
If it is called just once, then this hardly matters. But systemctl calls it 4 times per each failing unit, for example, and I don't think gcc can optimize the call away because it touches external state (tty, environment variables). So I think *unlikely* is appropriate.